### PR TITLE
[MM-32137] add config telemetry analyzers

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
The idea of this checker to create a `map[string]token.Pos` to have the reference for a struct field. Then, find out if the fields are used in the `telemetry` package.

e.g.
```golang
type Config struct{
	ServiceSettings ServiceSettings
}

type ServiceSettings struct {
	WebserverMode          *string
	SiteURL                *string    
}
```

The all possible `Config` fields are:
`[ServiceSettings.WebserverMode]token.Pos{}`
`[ServiceSettings.SiteURL]token.Pos{}`

Once we have the map, we check the `telemetry` package whether if they are used or not.

Currently we don't check if a config field is added to telemetry queue. I assumed that if a config field is used in the file, it probably added to the telemetry. This is a naive approach but I put my trust in our contributors.

We can add extra layer of check if a config field is supplied as a parameter into let's say `ts.SendTelemetry()` method. But we have to change the checker every time if we decide to change function name, or anything like that. Let me know if you think it would be better add this extra layer. (I have a branch for that)

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-32137

### CODE REVIEW

**Comment by agnivade** [comment]
Haven't taken a look at the code yet, but dare I ask, would it possible to have some tests on this? I am fully aware that this repo has 0 tests, but given the complexity of this, I think it would improve my confidence a bit during code review (and might as well catch some corner case bugs!).

File: telemetry/telemetry.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package telemetry
```
Comment (grundleborg) [inline telemetry/telemetry.go]: 0/5 but maybe we should give this analyser a more specific name, e.g. "config-telemetry" or something? Since that's all it does - it doesn't do anything else telemetry related.

File: telemetry/telemetry.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("could generate fields map: %w", err)
+	}
+
+	var allExpressions []string
+
+	// We look every expression in the file so that we can
+	// check if the expected references exist in this package.
+
+	// NOTE: this approach does not check if the reference (config field) is added
```
Comment (grundleborg) [inline telemetry/telemetry.go]: I *think* I'm comfortable with this approach, but 0/5 if others aren't.

A suggestion though, how about we move out the `trackConfig()` function into it's own file in the telemetry service? That way, the chance of the config field being used for something else in the file is much lower?

File: telemetry/telemetry.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+const (
+	telemetryPkgPath = "github.com/mattermost/mattermost-server/v5/services/telemetry"
+	modelPkgPath     = "github.com/mattermost/mattermost-server/v5/model"
+)
+
+// Analyzer describes telemetry status analysis function for a config setting
+var Analyzer = &analysis.Analyzer{
+	Name: "telemetry",
+	Doc:  "reports if a config field is not used in services/telemetry package",
```
Comment (grundleborg) [inline telemetry/telemetry.go]: Although this description is pedantically correct, I think it would be more useful if it actually conveyed the _intent_ of the analyser to someone reading it. Maybe this instead:

```suggestion
	Doc:  "reports config fields where telemetry is not sent (unless it is specifically indicated  that no telemetry will be sent for that field)",
```

**Review by grundleborg — CHANGES_REQUESTED** [review]
Some high level comments from me. I've looked through the code and it all makes sense to me, but I've close to zero experience with go-vet / AST stuff myself, so deferring to Agniva for the thorough review on the detail of the analyser implementation.

File: telemetry/telemetry.go (hunk 3 above)
Comment (isacikgoz) [reply telemetry/telemetry.go]: Sounds good to me.

File: telemetry/telemetry.go (hunk 2 above)
Comment (isacikgoz) [reply telemetry/telemetry.go]: It's doable, we can inspect a single file. Note that we also track plugins and we should also move things related to those as well.

**Review by grundleborg — APPROVED** [review]
LGTM. Over to Agniva for the detailed review of the analyzer ;)

File: configtelemetry/testdata/src/model/config.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	SERVICE_SETTINGS_DEFAULT_MAX_LOGIN_ATTEMPTS = 10
+	SERVICE_SETTINGS_DEFAULT_ALLOW_CORS_FROM    = ""
+	SERVICE_SETTINGS_DEFAULT_LISTEN_AND_ADDRESS = ":8065"
+	SERVICE_SETTINGS_DEFAULT_GFYCAT_API_KEY     = "2_KtH_W5"
+	SERVICE_SETTINGS_DEFAULT_GFYCAT_API_SECRET  = "3wLVZPiswc3DnaiaFoLkDvB4X0IV6CpMkj4tf2inJRsBY6-FnkT08zGmppWFgeof"
+
+	LOCAL_MODE_SOCKET_PATH = "/var/tmp/mattermost_local.socket"
+)
+
+type ServiceSettings struct {
```
Comment (agnivade) [inline configtelemetry/testdata/src/model/config.go:10]: I am not seeing lines like `// want ..`. Without those, I don't believe it is actually being tested.

File: configtelemetry/type.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"go/types"
+	"strings"
+)
+
+var ignoreComments = []string{
+	"telemetry: none",
+	"This field is deprecated and must not be used.",
```
Comment (agnivade) [inline configtelemetry/type.go]: Not needed for now, but a more idiomatic comment is "// Deprecated: do not use". We can do that in a future PR or even a HW ticket.

File: configtelemetry/type.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+// typeFieldMap will generate map of it's all possible selectors and adds them as a key. The value
```
Comment (agnivade) [inline configtelemetry/type.go]: Nit: s/it's/its

File: configtelemetry/telemetry.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/types"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+var (
```
Comment (agnivade) [inline configtelemetry/telemetry.go:18]: Any reason these can't be `const`?

File: configtelemetry/telemetry.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	if pass.Pkg.Path() != telemetryPkgPath {
+		return nil, nil
+	}
+
+	// we need to find model.Config definition to parse its Type Spec
+	var configObject *types.TypeName
+	for _, obj := range pass.TypesInfo.Uses {
+		if tn, ok := obj.(*types.TypeName); ok && obj.Name() == "Config" && obj.Pkg().Path() == modelPkgPath {
```
Comment (agnivade) [inline configtelemetry/telemetry.go]: We can move "Config" to be a const.

File: configtelemetry/type.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+// is the position of the declaration in the type spec.
+func typeFieldMap(fset *token.FileSet, obj *types.TypeName) (map[string]token.Pos, error) {
+	srcFile := fset.PositionFor(obj.Pos(), false).Filename
```
Comment (agnivade) [inline configtelemetry/type.go]: We should verify for valid positions first using `IsValid`

File: configtelemetry/type.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+	fileNode, err := parser.ParseFile(fset, srcFile, nil, parser.AllErrors|parser.ParseComments)
+	if err != nil {
+		return nil, fmt.Errorf("could parse %s: %w", srcFile, err)
+	}
+
+	var fieldsMap = make(map[string]token.Pos)
```
Comment (agnivade) [inline configtelemetry/type.go]: We are actually redundantly allocating here. The `fields` function anyways creates the map. We can remove the allocation from here.

File: configtelemetry/type.go [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+		expr := field.Type
+		if ptr, ok := field.Type.(*ast.StarExpr); ok {
+			expr = ptr.X
+		}
+
+		key := field.Names[0].Name
+
+		v, ok := expr.(*ast.Ident)
+		if !ok || v.Obj == nil || v.Obj.Decl == nil {
```
Comment (agnivade) [inline configtelemetry/type.go:67]: Wanted to understand this a bit better- can you explain when will v.Obj not be `nil`, and v.Obj.Decl be `nil`? Or is it just defence in depth?

File: configtelemetry/type.go [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+		if field.Comment != nil && ignored(field.Comment) {
+			continue
+		}
+
+		expr := field.Type
+		if ptr, ok := field.Type.(*ast.StarExpr); ok {
+			expr = ptr.X
+		}
+
+		key := field.Names[0].Name
```
Comment (agnivade) [inline configtelemetry/type.go:64]: Let's check the length first here.

File: configtelemetry/type.go [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+	var fieldsMap = make(map[string]token.Pos)
+
+	ast.Inspect(fileNode, func(n ast.Node) bool {
+		decl, ok := n.(*ast.GenDecl)
+		if !ok || decl.Tok != token.TYPE {
+			return true
+		}
+
+		spec, ok := decl.Specs[0].(*ast.TypeSpec)
```
Comment (agnivade) [inline configtelemetry/type.go:29]: Check length first.

File: configtelemetry/telemetry.go [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	configMap, err := typeFieldMap(pass.Fset, configObject)
+	if err != nil {
+		return nil, fmt.Errorf("could generate fields map: %w", err)
+	}
+
+	var allExpressions []string
+
+	// We look every expression in the file so that we can
```
Comment (agnivade) [inline configtelemetry/telemetry.go]: We look _at_

File: configtelemetry/telemetry.go [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+	// check if the expected references exist in this package.
+
+	// NOTE: this approach does not check if the reference (config field) is added
+	// to telemetry queue, it assumes that if a config field is used in the file,
+	// it probably added to the telemetry. This is a naive approach but should work.
+	// Also, we can narrow down this for specific function calls, declarations etc.
```
Comment (agnivade) [inline configtelemetry/telemetry.go:79]: I am wondering how hard would it be to walk up a node and verify that the expression is used as a map value? Or even verify that all such expressions are only used inside `ts.sendTelemetry` method.

Just checking for these expressions within a package seems very broad to me. Would be good to have something more robust if possible.

File: configtelemetry/testdata/src/model/config.go (hunk 4 above)
Comment (isacikgoz) [reply configtelemetry/testdata/src/model/config.go:10]: Discussed offline, it was not possible to read `//want ..` comments from another package (`model`) since the `anaylsis.Pass` has files for `telemetry` package only. With a workaround we add `model/config.go` into `analysis.Pass` for the telemetry package. Will update accordingly.

File: configtelemetry/type.go (hunk 5 above)
Comment (isacikgoz) [reply configtelemetry/type.go:17]: Sure, but I can do it since we are here. Not a big deal :)

File: configtelemetry/type.go (hunk 11 above)
Comment (isacikgoz) [reply configtelemetry/type.go:67]: It's actually to be defensive.

File: configtelemetry/telemetry.go (hunk 15 above)
Comment (isacikgoz) [reply configtelemetry/telemetry.go:79]: There are some occasions where config values are not directly used in `ts.sendTelemetry` (e.g. `PluginSettings`) so it gets quickly complicated to track those. I think checking if the expression is used as a map value won't be that hard.

Even we check if we the values are used as map value or in `sendTelemetry` method arg, it's still possible to trick the checker. At some point we need to trust the developer. I can add the map-value layer if you want. I'm 0/5 on either approach.

File: configtelemetry/type.go (hunk 5 above)
Comment (isacikgoz) [reply configtelemetry/type.go]: Not a big deal of work, we can do it since we are here 👍

File: configtelemetry/telemetry.go (hunk 7 above)
Comment (isacikgoz) [reply configtelemetry/telemetry.go:18]: To change their values while testing. Since they are unexported I think it wouldn't hurt that much.

File: configtelemetry/telemetry.go (hunk 15 above)
Comment (agnivade) [reply configtelemetry/telemetry.go:79]: Gave it some more thought. I agree with you. However, let's add an escape hatch as a custom comment to ignore a config field to let a developer use it without letting the vet pass think of it as something to be used in telemetry.

**Review by agnivade — APPROVED** [review]
Looks great!
